### ISSUE TITLE
Version declared in a python_require for the consumer is now an error: regression in 1.11.0

### ISSUE THREAD

**Issue by solvingj** [conan-io/conan#4355]
**Description:** 
After upgrading to Conan 1.11.0 or higher, our recipes which do not directly define their version number now produce the error below.  I don't know if there are other scenarios in which the `version` number can be defined somewhere other than the `conanfile.py`, but if there are, it seems likely they are also affected. 

**Workaround:**
`pip install conan==1.10.1`

**Steps to reproduce:**
Simply try to build any recipe that doesn't have a version, and which gets in version via `python_requires`.  
 
Here is an example:
`https://github.com/bincrafters/conan-boost_callable_traits`

Here is the output: 
```
>conan create . bincrafters/testing

ERROR: conanfile didn't specify version
```

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by SSE4** [comment]
some investigation details.
the code in version `1.10.0` (`conans\client\loader.py`):
```
        if not conanfile.version:
            if version:
                conanfile.version = version
            else:
                raise ConanException("conanfile didn't specify version")
```
and in `1.11.2`:
```
        if "version" in conanfile.__dict__:
            if version and version != conanfile.version:
                raise ConanException("Package recipe exported with version %s!=%s"
                                     % (version, conanfile.version))
            else:
                version = conanfile.version
```
so, previously it used just `conanfile.version`, now it uses `"version" in conanfile.__dict__`, which are not the same.
`__dict__` appears to be some python internal magic (https://docs.python.org/3.7/library/stdtypes.html#object.__dict__):
```
A dictionary or other mapping object used to store an object’s (writable) attributes.
```
and it doesn't play very well with inheritance (https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy):
```
A class has a namespace implemented by a dictionary object. Class attribute references are translated to lookups in this dictionary, e.g., C.x is translated to C.__dict__["x"] (although there are a number of hooks which allow for other means of locating attributes). When the attribute name is not found there, the attribute search continues in the base classes
```
I have the following simple sandbox to confirm the theory:
```
from __future__ import print_function

class A(object):
    version = "1.0"

class B(A):
    pass

print(A.version)
print(B.version)
print("version" in A.__dict__)
print("version" in B.__dict__)
```
and the output is:
```
1.0
1.0
True
False
```
so, `1.11.2` clearly breaks conanfile inheritance use-case (inheritance in general, not just an experimental `python_requires` feature)

/cc @conan-io/barbarians @uilianries

**Comment by solvingj** [comment]
thanks for the speedy investigation!

**Comment by solvingj** [comment]
That makes sense.  I knew graph lock would cause a few difficult decisions, and I am sure it was the right thing to do. 

Of note, I know that there are at least a few enterprise organizations who have been using inheritance long before `python_requires` came around.  They use custom python packages to distribute their base classes and other wrapper functions via `pip`, along with `conan`, and just `import` their base classes.  In various slack discussions, this has been discouraged by the Conan team and Bincrafters for OSS packages, because it would introduce a `pip` dependency outside of Conan and that's unacceptable, since Conan can be distributed as a standalone installer.  However, this is a very attractive thing to do for enterprise codebases, and I am unware of any general Conan documentation explicitly stating that this practice was discouraged or forbidden.  

In summary, I think that at some point, as these organizations reach their upgrade cycles and try to move up to Conan > 1.10.0, you're going to get a lot of feedback that these users would consider this major breaking change for those use cases, and it should have marked 2.0.  Our team will be able to adjust our strategy based on the two links you provided, so I'm closing ticket.
